connectorx/src/sources/postgres/mod.rs

```diff
@@ -961,3 +966,771 @@ impl_produce!(
+    fn parser(&mut self) -> Self::Parser<'_> {
+        let rows = self.conn.simple_query(self.query.as_str())?; // unless reading the data, it seems like issue the query is fast
+        PostgresSimpleSourceParser::new(rows, &self.schema)
+    }
+
+    fn nrows(&self) -> usize {
+        self.nrows
+    }
+
+    fn ncols(&self) -> usize {
+        self.ncols
+    }
+}
+
+pub struct PostgresSimpleSourceParser {
+    rows: Vec<SimpleQueryMessage>,
+    ncols: usize,
+    current_col: usize,
+    current_row: usize,
+}
+impl<'a> PostgresSimpleSourceParser {
+    pub fn new(rows: Vec<SimpleQueryMessage>, schema: &[PostgresTypeSystem]) -> Self {
+        Self {
+            rows,
+            ncols: schema.len(),
+            current_row: 0,
+            current_col: 0,
+        }
+    }
+
+    #[throws(PostgresSourceError)]
+    fn next_loc(&mut self) -> (usize, usize) {
+        let ret = (self.current_row, self.current_col);
+        self.current_row += (self.current_col + 1) / self.ncols;
+        self.current_col = (self.current_col + 1) % self.ncols;
+        ret
+    }
+}
+
+impl<'a> PartitionParser<'a> for PostgresSimpleSourceParser {
+    type TypeSystem = PostgresTypeSystem;
+    type Error = PostgresSourceError;
+
+    #[throws(PostgresSourceError)]
+    fn fetch_next(&mut self) -> (usize, bool) {
+        self.current_row = 0;
+        self.current_col = 0;
+        (self.rows.len() - 1, true) // last message is command complete
+    }
+}
+
+macro_rules! impl_simple_produce_unimplemented {
+    ($($t: ty,)+) => {
+        $(
+            impl<'r, 'a> Produce<'r, $t> for PostgresSimpleSourceParser {
+                type Error = PostgresSourceError;
+
+                #[throws(PostgresSourceError)]
+                fn produce(&'r mut self) -> $t {
+                   unimplemented!("not implemented!");
+                }
+            }
+
+            impl<'r, 'a> Produce<'r, Option<$t>> for PostgresSimpleSourceParser {
+                type Error = PostgresSourceError;
+
+                #[throws(PostgresSourceError)]
+                fn produce(&'r mut self) -> Option<$t> {
+                   unimplemented!("not implemented!");
+                }
+            }
+        )+
+    };
+}
+
+macro_rules! impl_simple_produce {
+    ($($t: ty,)+) => {
+        $(
+            impl<'r> Produce<'r, $t> for PostgresSimpleSourceParser {
+                type Error = PostgresSourceError;
+
+                #[throws(PostgresSourceError)]
+                fn produce(&'r mut self) -> $t {
+                    let (ridx, cidx) = self.next_loc()?;
+                    let val = match &self.rows[ridx] {
+                        SimpleQueryMessage::Row(row) => match row.try_get(cidx)? {
+                            Some(s) => s
+                                .parse()
+                                .map_err(|_| ConnectorXError::cannot_produce::<$t>(Some(s.into())))?,
+                            None => panic!("get NULL for non-NULL column"),
```

Comment:
Would be better if we can throw an error instead of panic in the code (unless it's for our own testing). Can you try to throw an `cannot_produce` or `anyhow` error for all the panics?

Reply:
it was for debugging, will do this shortly :)